File: dmo-client-main/DProject/EffectMng.cpp

#include "stdafx.h"
#include "EffectMng.h"

CEffectMng*		g_pEffectMng = NULL;

CEffectMng::CEffectMng()
{

}

CEffectMng::~CEffectMng()
{

}

void CEffectMng::GlobalInit()
{
	assert_cs( g_pEffectMng == NULL );
	g_pEffectMng = NiNew CEffectMng;

	sCREATEINFO prop;
	prop.s_dwPropPlag = PR_EFFECT_ENABLE;
	prop.s_dwOptionPlag = 0;
	prop.s_eInsertGeomType = CGeometry::Normal;

	g_pEffectMng->Init( 0, 0, NiPoint3::ZERO, 0.0f, &prop );
	g_pEffectMng->Init_Marble();
}

void CEffectMng::GlobalShotDown()
{
	if( g_pEffectMng == NULL )
		return;

	g_pEffectMng->Destroy_Marble();
	g_pEffectMng->Delete();
	NiDelete g_pEffectMng;
	g_pEffectMng = NULL;
}

void CEffectMng::Reset()
{
	GetProp_Effect()->Reset();
	_ResetMarble();
}

CsC_EffectObject* CEffectMng::AddEffect( char* cEffectPath, NiPoint3 vPos, float fScale )
{
	return GetProp_Effect()->AddEffect( cEffectPath, vPos, fScale );
}

void CEffectMng::DeleteEffect( CsC_AvObject* pDelEffect )
{
	GetProp_Effect()->DeleteEffect( pDelEffect );
}

void CEffectMng::Update( float fDeltaTime )
{
	GetProp_Effect()->PreUpdate( fDeltaTime );
	_Update_Marble( fDeltaTime );
}

void CEffectMng::Render()
{
	GetProp_Effect()->Render();
	_Render_Marble();
}

//================================================================================================================
//
//		포인트 구슬
//
//================================================================================================================

void CEffectMng::Init_Marble()
{
#define MARBLE_EVOLUTION_OBJECT_PATH	"Data\\EtcObject\\Marble_Evo.nif"

	NiStream kStream;
	if( !kStream.Load( MARBLE_EVOLUTION_OBJECT_PATH ) )
	{
		//CsMessageBoxA( MB_OK, "%s 경로\n에서 오브젝트를 찾지 못했습니다.", MARBLE_EVOLUTION_OBJECT_PATH );
		return;
	}

	NiNodePtr pNode = (NiNode*)kStream.GetObjectAt(0);
	nsCSGBFUNC::InitAnimation( pNode, NiTimeController::APP_INIT, NiTimeController::LOOP );
	m_MarbleObj[ MT_EVOLUTION ].SetNiObject( pNode, CGeometry::Normal );
}

void CEffectMng::Destroy_Marble()
{
	_ResetMarble();

	sMARBLE_INFO* pMarble = NULL;
	for( int i=0; i<MT_MAX_CNT; ++i )
	{
		while( m_mpMarble[ i ].empty() == false )
		{
			pMarble = m_mpMarble[ i ].front();
			pMarble->s_NodeObj.Delete();
			delete pMarble;
			m_mpMarble[ i ].pop();
		}
	}	

	for( int i=0; i<MT_MAX_CNT; ++i )
		m_MarbleObj[ i ].Delete();
}

void CEffectMng::AddEffect_Marble( eMARBLE_TYPE mt, NiPoint3 vSrc, CsC_AvObject* pOwner, int nCnt )
{
	for( int i=0; i<nCnt; ++i )
	{
		sMARBLE_INFO* pNew = _NewMarble( mt, vSrc, pOwner );
		m_listMarble.push_back( pNew );
	}
}

void CEffectMng::_Update_Marble( float fDeltaTime )
{
	sMARBLE_INFO* pMarble = NULL;
	std::list< sMARBLE_INFO* >::iterator it = m_listMarble.begin();
	std::list< sMARBLE_INFO* >::iterator itEnd = m_listMarble.end();
	while( it!=itEnd )
	{
		pMarble = *it;
		pMarble->s_fAniTime += fDeltaTime;

		switch( pMarble->s_eStep )
		{
		case sMARBLE_INFO::CREATE:
			{
				float fRate = pMarble->s_fAniTime*2.4f;
				pMarble->s_vCurPos = nsCSGBFUNC::CalMullRom(&pMarble->s_vSpline1,
															&pMarble->s_vInitPos,
															&pMarble->s_vTrPos,
															&pMarble->s_vSpline2,
															CsMin( 1.0f, fRate ) );

				if( fRate > 1.0f )
				{
					/*nsCSGBFUNC::StartAnimation( pMarble->s_NodeObj.m_pNiNode, pMarble->s_fAniTime );
					pMarble->s_eStep = sMARBLE_INFO::ON_TR;
					pMarble->s_fOnTrTime += pMarble->s_fAniTime;*/

					pMarble->s_eStep = sMARBLE_INFO::FALLOW;
					pMarble->s_fAniTime = 0.0f;
				}

				pMarble->s_NodeObj.m_pNiNode->SetTranslate( pMarble->s_vCurPos );
			}
			break;
		/*case sMARBLE_INFO::ON_TR:
			{
				if( pMarble->s_fAniTime > pMarble->s_fOnTrTime )
				{
					nsCSGBFUNC::StopAnimation( pMarble->s_NodeObj.m_pNiNode );
					pMarble->s_eStep = sMARBLE_INFO::FALLOW;
					pMarble->s_fAniTime = 0.0f;
				}
			}
			break;*/
		case sMARBLE_INFO::FALLOW:
			{
				bool bEnd = false;
				if( pMarble->s_pOwner->IsLoad() )
				{
					// Dir
					NiPoint3 vDest = pMarble->s_pOwner->GetPos();
					vDest.z += pMarble->s_pOwner->GetToolHeight()*0.5f;
					NiPoint3 vSrc = pMarble->s_vCurPos;
					NiPoint3 vDir = vDest - vSrc;
					float fDist = vDir.Unitize();
					// Velocity
					float fSpeed = pow( pMarble->s_fAniTime+1.0f, 6 )*60.0f*g_fDeltaTime + 50.0f;
					//float fSpeed = pow( pMarble->s_fAniTime+1.0f, 4 )*g_fDeltaTime;
					if( fDist < fSpeed )
					{
						fSpeed = fDist;
						bEnd = true;
					}
					pMarble->s_vCurPos = vSrc + vDir*fSpeed;
					pMarble->s_NodeObj.m_pNiNode->SetTranslate( pMarble->s_vCurPos );
				}
				else
				{
					bEnd = true;
				}

				if( bEnd == true )
				{
					pMarble->s_eStep		= sMARBLE_INFO::DIE;
					pMarble->s_fOnTrTime	= 0.2f;					
				}
			}
			break;
		case sMARBLE_INFO::DIE:
			{
				bool bEnd = false;
				if( pMarble->s_pOwner->IsLoad() )
				{
					if( ( pMarble->s_fOnTrTime -= g_fDeltaTime ) > 0 )
					{
						NiPoint3 vDest = pMarble->s_pOwner->GetPos();
						vDest.z += pMarble->s_pOwner->GetToolHeight()*0.5f;
						pMarble->s_NodeObj.m_pNiNode->SetTranslate( vDest );
					}
					else
					{
						bEnd = true;
					}
				}
				else
				{
					bEnd = true;
				}

				if( bEnd == true )
				{
					it = m_listMarble.erase( it );
					_DelMarble( pMarble );
					continue;
				}
			}
			break;
		default:
			assert_cs( false );
		}		
		pMarble->s_NodeObj.m_pNiNode->Update( pMarble->s_fAniTime );
		++it;
	}
}

void CEffectMng::_Render_Marble()
{
	sMARBLE_INFO* pMarble = NULL;
	std::list< sMARBLE_INFO* >::iterator it = m_listMarble.begin();
	std::list< sMARBLE_INFO* >::iterator itEnd = m_listMarble.end();
	for( ; it!=itEnd; ++it )
	{
		pMarble = *it;
		if( CAMERA_ST.UserCulling( pMarble->s_NodeObj.m_pNiNode ) == true )
		{
			pMarble->s_NodeObj.RenderAbsolute();
		}
	}
}

CEffectMng::sMARBLE_INFO* CEffectMng::_NewMarble( eMARBLE_TYPE mt, NiPoint3 vSrc, CsC_AvObject* pOwner )
{
	sMARBLE_INFO* pMarble = NULL;
	if( m_mpMarble[ mt ].empty() == true )
	{
		pMarble = csnew sMARBLE_INFO;
		pMarble->s_NodeObj.SetNiObject( (NiNode*)m_MarbleObj[ mt ].m_pNiNode->Clone(), CGeometry::Normal );
	}
	else
	{
		pMarble = m_mpMarble[ mt ].front();
		m_mpMarble[ mt ].pop();
	}

	pMarble->s_fAniTime = 0;
	pMarble->s_eType = mt;
	pMarble->s_eStep = sMARBLE_INFO::CREATE;
	pMarble->s_pOwner = pOwner;
	pMarble->s_vInitPos = vSrc;

	// 각
	NiPoint3 v = pOwner->GetPos() - vSrc;
	v.z = 0.0f;
	v.Unitize();
	float fRot = NiACos( NiPoint3( 0.0f, -1.0f, 0.0f ).Dot( v ) );
	if( NiPoint3( 0.0f, -1.0f, 0.0f ).Cross( v ).z > 0 )
		fRot = -fRot;	

	NiMatrix3 mat;
	int nRes = nBase::rand()%120 - 60;
	mat.MakeZRotation( fRot + CsD2R( nRes ) );
	NiPoint3 vDelta = mat * NiPoint3( 0, 1, 0 );

	// 거리
	vDelta *= (float)nBase::rand( 400, 600 );
	// 랜덤 위치
	pMarble->s_vTrPos.x = vSrc.x + vDelta.x;
	pMarble->s_vTrPos.y = vSrc.y + vDelta.y;
	pMarble->s_vTrPos.z = nsCsGBTerrain::g_pCurRoot->GetHeight( pMarble->s_vTrPos ) + (float)nBase::rand( 50, 210 );

	// 스플라인 위치
	pMarble->s_vSpline1 = vSrc + NiPoint3( 0, 0, -100 );
	pMarble->s_vSpline2 = pMarble->s_vTrPos + NiPoint3( 0, 0, -(float)nBase::rand( 1000, 3000 ) );

	// 지형 존재 시간
	pMarble->s_fOnTrTime = nBase::rand( 500, 1000 )*0.001f;

	return pMarble;
}

void CEffectMng::_ResetMarble()
{
	std::list< sMARBLE_INFO* >::iterator it = m_listMarble.begin();
	std::list< sMARBLE_INFO* >::iterator itEnd = m_listMarble.end();
	for( ; it!=itEnd; ++it )
	{
		_DelMarble( *it );
	}
	m_listMarble.clear();
}

void CEffectMng::_DelMarble( sMARBLE_INFO* pInfo )
{
	m_mpMarble[ pInfo->s_eType ].push( pInfo );
}

File: dmo-client-main/DProject.CSharp/Game/EffectMng.cs
using DProject.Gamebryo;
using DProject.Runtime;

namespace DProject.Game;

public sealed class CsC_EffectObject : CsC_AvObject
{
    public string EffectPath { get; private set; } = string.Empty;
    public float Scale { get; private set; } = 1.0f;
    public float LifeTime { get; private set; }

    public void InitEffect(uint uid, string effectPath, NiPoint3 position, float scale)
    {
        EffectPath = effectPath;
        Scale = scale;
        LifeTime = 0.0f;
        Init(uid, uid & 0x3FFF, 0, 0, position, 0.0f, effectPath);
    }

    public override void Update(float deltaTime, bool animation = true)
    {
        LifeTime += deltaTime;
    }
}

public readonly record struct EffectRenderCommand(uint Uid, string EffectPath, NiPoint3 Position, float Scale);

public sealed class CEffectMng : CsC_AvObject
{
    private readonly List<CsC_EffectObject> _effects = [];
    private readonly List<sMARBLE_INFO> _marbles = [];
    private readonly Queue<sMARBLE_INFO>[] _marblePool;
    private uint _nextEffectUid = 1;

    public CEffectMng()
    {
        _marblePool = Enumerable.Range(0, (int)eMARBLE_TYPE.MT_MAX_CNT)
            .Select(static _ => new Queue<sMARBLE_INFO>())
            .ToArray();
    }

    public enum eMARBLE_TYPE
    {
        MT_EVOLUTION,
        MT_MAX_CNT
    }

    public IReadOnlyList<CsC_EffectObject> Effects => _effects;
    public IReadOnlyList<sMARBLE_INFO> Marbles => _marbles;

    public void Reset()
    {
        foreach (CsC_EffectObject effect in _effects)
        {
            effect.Delete();
        }

        _effects.Clear();
        ResetMarble();
    }

    public override void Update(float fDeltaTime, bool animation = true)
    {
        foreach (CsC_EffectObject effect in _effects)
        {
            effect.Update(fDeltaTime, animation);
        }

        UpdateMarble(fDeltaTime);
    }

    public void Render(RenderCommandQueue? renderCommands = null)
    {
        foreach (CsC_EffectObject effect in _effects)
        {
            EffectRenderCommand command = new(effect.GetUniqID(), effect.EffectPath, effect.GetPos(), effect.Scale);
            renderCommands?.Enqueue(() =>
            {
                _ = command;
                // Placeholder until CsC_EffectObject/Gamebryo rendering is ported.
            });
        }

        RenderMarble(renderCommands);
    }

    public CsC_EffectObject AddEffect(string cEffectPath, NiPoint3 vPos, float fScale)
    {
        var effect = new CsC_EffectObject();
        effect.InitEffect(_nextEffectUid++, cEffectPath, vPos, fScale);
        _effects.Add(effect);
        return effect;
    }

    public void DeleteEffect(CsC_AvObject pDelEffect)
    {
        _effects.RemoveAll(effect => effect.GetUniqID() == pDelEffect.GetUniqID());
        pDelEffect.Delete();
    }

    public void Init_Marble()
    {
    }

    public void Destroy_Marble()
    {
        ResetMarble();
        foreach (Queue<sMARBLE_INFO> pool in _marblePool)
        {
            pool.Clear();
        }
    }

    public void AddEffect_Marble(eMARBLE_TYPE mt, NiPoint3 vSrc, CsC_AvObject pOwner, int nCnt)
    {
        for (int i = 0; i < nCnt; i++)
        {
            _marbles.Add(NewMarble(mt, vSrc, pOwner));
        }
    }

    private void UpdateMarble(float fDeltaTime)
    {
        for (int i = _marbles.Count - 1; i >= 0; i--)
        {
            sMARBLE_INFO marble = _marbles[i];
            marble.s_fAniTime += fDeltaTime;

            switch (marble.s_eStep)
            {
                case sMARBLE_INFO.eSTEP.CREATE:
                    UpdateMarbleCreate(marble);
                    break;
                case sMARBLE_INFO.eSTEP.FALLOW:
                    UpdateMarbleFollow(marble, fDeltaTime);
                    break;
                case sMARBLE_INFO.eSTEP.DIE:
                    if (UpdateMarbleDie(marble, fDeltaTime))
                    {
                        _marbles.RemoveAt(i);
                        DelMarble(marble);
                    }
                    break;
            }
        }
    }

    private static void UpdateMarbleCreate(sMARBLE_INFO marble)
    {
        float rate = MathF.Min(1.0f, marble.s_fAniTime * 2.4f);
        marble.s_vCurPos = Lerp(marble.s_vInitPos, marble.s_vTrPos, rate);
        if (rate >= 1.0f)
        {
            marble.s_eStep = sMARBLE_INFO.eSTEP.FALLOW;
            marble.s_fAniTime = 0.0f;
        }
    }

    private static void UpdateMarbleFollow(sMARBLE_INFO marble, float deltaTime)
    {
        if (!marble.s_pOwner.IsLoad())
        {
            marble.s_eStep = sMARBLE_INFO.eSTEP.DIE;
            marble.s_fOnTrTime = 0.2f;
            return;
        }

        NiPoint3 dest = marble.s_pOwner.GetPos().WithZ(marble.s_pOwner.GetPos().Z + (marble.s_pOwner.GetToolHeight() * 0.5f));
        NiPoint3 delta = dest - marble.s_vCurPos;
        float distance = delta.Length();
        if (distance <= 0.001f)
        {
            marble.s_eStep = sMARBLE_INFO.eSTEP.DIE;
            marble.s_fOnTrTime = 0.2f;
            return;
        }

        float speed = MathF.Pow(marble.s_fAniTime + 1.0f, 6.0f) * 60.0f * deltaTime + 50.0f;
        if (distance <= speed)
        {
            marble.s_vCurPos = dest;
            marble.s_eStep = sMARBLE_INFO.eSTEP.DIE;
            marble.s_fOnTrTime = 0.2f;
            return;
        }

        marble.s_vCurPos += delta.Unitize() * speed;
    }

    private static bool UpdateMarbleDie(sMARBLE_INFO marble, float deltaTime)
    {
        if (!marble.s_pOwner.IsLoad())
        {
            return true;
        }

        marble.s_fOnTrTime -= deltaTime;
        NiPoint3 ownerPos = marble.s_pOwner.GetPos();
        marble.s_vCurPos = ownerPos.WithZ(ownerPos.Z + (marble.s_pOwner.GetToolHeight() * 0.5f));
        return marble.s_fOnTrTime <= 0.0f;
    }

    private void RenderMarble(RenderCommandQueue? renderCommands)
    {
        foreach (sMARBLE_INFO marble in _marbles)
        {
            EffectRenderCommand command = new((uint)marble.GetHashCode(), marble.s_eType.ToString(), marble.s_vCurPos, 1.0f);
            renderCommands?.Enqueue(() =>
            {
                _ = command;
                // Placeholder until marble NiNode rendering is ported.
            });
        }
    }

    private sMARBLE_INFO NewMarble(eMARBLE_TYPE mt, NiPoint3 vSrc, CsC_AvObject pOwner)
    {
        Queue<sMARBLE_INFO> pool = _marblePool[(int)mt];
        sMARBLE_INFO marble = pool.Count > 0 ? pool.Dequeue() : new sMARBLE_INFO();
        marble.s_eType = mt;
        marble.s_eStep = sMARBLE_INFO.eSTEP.CREATE;
        marble.s_pOwner = pOwner;
        marble.s_fAniTime = 0.0f;
        marble.s_fOnTrTime = Random.Shared.Next(500, 1000) * 0.001f;
        marble.s_vInitPos = vSrc;
        marble.s_vTrPos = PickTransitionPoint(vSrc, pOwner.GetPos());
        marble.s_vSpline1 = vSrc + new NiPoint3(0.0f, 0.0f, -100.0f);
        marble.s_vSpline2 = marble.s_vTrPos + new NiPoint3(0.0f, 0.0f, -Random.Shared.Next(1000, 3000));
        marble.s_vCurPos = vSrc;
        return marble;
    }

    private void ResetMarble()
    {
        foreach (sMARBLE_INFO marble in _marbles)
        {
            DelMarble(marble);
        }

        _marbles.Clear();
    }

    private void DelMarble(sMARBLE_INFO pInfo)
    {
        _marblePool[(int)pInfo.s_eType].Enqueue(pInfo);
    }

    private static NiPoint3 PickTransitionPoint(NiPoint3 source, NiPoint3 owner)
    {
        NiPoint3 direction = (owner - source).WithZ(0.0f).Unitize();
        if (direction == NiPoint3.Zero)
        {
            direction = new NiPoint3(0.0f, 1.0f, 0.0f);
        }

        float angle = Random.Shared.Next(-60, 60) * (MathF.PI / 180.0f);
        float cos = MathF.Cos(angle);
        float sin = MathF.Sin(angle);
        NiPoint3 rotated = new(
            (direction.X * cos) - (direction.Y * sin),
            (direction.X * sin) + (direction.Y * cos),
            0.0f);
        float distance = Random.Shared.Next(400, 600);
        return source + (rotated * distance) + new NiPoint3(0.0f, 0.0f, Random.Shared.Next(50, 210));
    }

    private static NiPoint3 Lerp(NiPoint3 start, NiPoint3 end, float rate)
    {
        return start + ((end - start) * rate);
    }

    public sealed class sMARBLE_INFO
    {
        public enum eSTEP
        {
            CREATE,
            ON_TR,
            FALLOW,
            DIE
        }

        public eMARBLE_TYPE s_eType;
        public eSTEP s_eStep;
        public CsC_AvObject s_pOwner = new();
        public float s_fAniTime;
        public float s_fOnTrTime;
        public NiPoint3 s_vInitPos;
        public NiPoint3 s_vTrPos;
        public NiPoint3 s_vSpline1;
        public NiPoint3 s_vSpline2;
        public NiPoint3 s_vCurPos;
    }
}
